test-data/unit/check-classes.test

```diff
@@ -6891,3 +6891,71 @@ class B(A):  # E: Final class __main__.B has abstract attributes "foo"
 [case testUndefinedBaseclassInNestedClass]
 class C:
     class C1(XX): pass  # E: Name "XX" is not defined
+
+[case testNoReturn]
+from typing import NoReturn
+
+class A:
+    def __new__(cls) -> NoReturn: ...
+
+class B:
+    def __init__(self) -> NoReturn: ...
+
+class C:
+    def __new__(cls) -> "C": ...
+    def __init__(self) -> NoReturn: ...
+
+class D:
+    def __new__(cls) -> NoReturn: ...
+    def __init__(self) -> NoReturn: ...
```

Comment:
``` python
class E:
    def __new__(cls) -> NoReturn: ...
    def __init__(self) -> None: ...
```
Notable in it's absence is the example above. Unfortunately this example fails to properly recognize `NoReturn` when instantiating the class, as mypy tends to ignore `__new__` if `__init__` is defined:
https://github.com/python/mypy/blob/8e82171a3e68a5180fab267cad4d2b7cfa1f5cdc/mypy/checkmember.py#L969-L973